### PR TITLE
Add capability to use remote plugins via gRPC

### PR DESCRIPTION
**Pull request by tomasol**
Summary:
Add writer registry
Add gRPC bindings to allow using of remote plugins. Support for following RPCs has been added: GetCapabilities, Create, Read, Update, Delete, i.e. plugin discovery, ReaderPlugins, WriterPlugins.
Sample configuration added to doc/config-sample.

Writer registry contains commented code, since some functionality, that is not required for remote plugins, will be available later. This will be fixed in next PRs.

### CODE REVIEW

File: devmand/gateway/CMakeLists.txt [hunk 1]
```diff
@@ -158,10 +158,31 @@ add_library(devman_channel_cli
   ${PROJECT_SOURCE_DIR}/src/devmand/channels/cli/datastore/BindingAwareDatastoreTransaction.h
   ${PROJECT_SOURCE_DIR}/src/devmand/channels/cli/datastore/DatastoreDiff.h
   ${PROJECT_SOURCE_DIR}/src/devmand/channels/cli/datastore/BindingAwareDatastore.cpp
-  ${PROJECT_SOURCE_DIR}/src/devmand/channels/cli/datastore/BindingAwareDatastore.h)
+  ${PROJECT_SOURCE_DIR}/src/devmand/channels/cli/datastore/BindingAwareDatastore.h
+  ${PROJECT_SOURCE_DIR}/src/devmand/channels/cli/plugin/protocpp/Common.pb.cpp
+  ${PROJECT_SOURCE_DIR}/src/devmand/channels/cli/plugin/protocpp/Common.grpc.pb.cpp
+  ${PROJECT_SOURCE_DIR}/src/devmand/channels/cli/plugin/protocpp/ReaderPlugin.pb.cpp
```
Comment (marosmars) [inline devmand/gateway/CMakeLists.txt:164]: I would move framework and plugins related code to devmand/devices/cli

File: devmand/gateway/src/devmand/channels/cli/engine/Engine.cpp [hunk 2]
```diff
@@ -61,16 +68,68 @@ void Engine::initLogging(uint32_t verbosity, bool callInitMlog) {
...
+  }
+  return move(pReg);
+}
+
 /*
  * Keep alive cli layer needs separate executor for protecting against resource
  * starvation - connection failures should always be detected.
  * SSH cli (SshSessionAsync) - separate executor for ssh layer.
  * All other layers share common threadpool executor.
+ * Plugin executor is used used for remote plugins.
```
Comment (marosmars) [inline devmand/gateway/src/devmand/channels/cli/engine/Engine.cpp]: used used

File: devmand/gateway/src/devmand/channels/cli/engine/Engine.cpp [hunk 3]
```diff
@@ -114,6 +177,28 @@ shared_ptr<ModelRegistry> Engine::getModelRegistry() const {
   return mreg;
 }
 
+shared_ptr<DeviceContext> Engine::getDeviceContext(const DeviceType& type) {
+  return pluginRegistry->getDeviceContext(type);
+}
+
+// TODO should this be singleton/cached?
```
Comment (marosmars) [inline devmand/gateway/src/devmand/channels/cli/engine/Engine.cpp]: Cached, yes per device type. Constructing ReaderRegistry / WriterRegistry is expensive

File: devmand/gateway/src/devmand/channels/cli/engine/Engine.cpp [hunk 4]
```diff
@@ -61,16 +68,68 @@ void Engine::initLogging(uint32_t verbosity, bool callInitMlog) {
...
+    shared_ptr<Executor> executor) {
+  unique_ptr<PluginRegistry> pReg = make_unique<PluginRegistry>();
+  // all local plugins should be added here:
+  pReg->registerPlugin(make_shared<UbntInterfacePlugin>(
+      modelRegistry->getBindingContext(Model::OPENCONFIG_2_4_3)));
+  pReg->registerPlugin(make_shared<UbntStpPlugin>(
+      modelRegistry->getBindingContext(Model::OPENCONFIG_2_4_3)));
+  pReg->registerPlugin(make_shared<UbntNetworksPlugin>(
+      modelRegistry->getBindingContext(Model::OPENCONFIG_2_4_3)));
+  // TODO read configuration, add remote plugins
```
Comment (marosmars) [inline devmand/gateway/src/devmand/channels/cli/engine/Engine.cpp]: this TODO seems done

File: devmand/gateway/src/devmand/channels/cli/engine/Engine.cpp (hunk 3 above)
Comment (marosmars) [reply devmand/gateway/src/devmand/channels/cli/engine/Engine.cpp]: Doesn't have to be done now

File: devmand/gateway/src/devmand/magma/DevConf.cpp [hunk 5]
```diff
@@ -21,12 +21,42 @@ static constexpr const char* ymlExt = ".yml";
...
 static constexpr const char* mconfigExt = ".mconfig";
 
+static folly::dynamic parsePluginConfigFromYaml(
+    const std::string& deviceConfigurationFile) {
+  try {
+    YAML::Node devicesFile = YAML::LoadFile(deviceConfigurationFile);
+    const YAML::Node& pluginConfigYamlNode = devicesFile["pluginConfig"];
+    if (pluginConfigYamlNode.IsScalar()) {
+      const string& filename = pluginConfigYamlNode.as<std::string>();
+      LOG(INFO) << "Loading plugin config file: " << filename;
```
Comment (marosmars) [inline devmand/gateway/src/devmand/magma/DevConf.cpp:31]: MLOG ?

**Review by marosmars — APPROVED** [review]
Looks good. A couple smaller comments (can be addressed later).

Also, we should rename StructuredUbntDevice to StructuredDevice or similar since this is no longer just Ubiquiti

File: devmand/gateway/src/devmand/magma/DevConf.cpp (hunk 5 above)
Comment (tomasol) [reply devmand/gateway/src/devmand/magma/DevConf.cpp:31]: I kept it consistent within DevConf.cpp

File: devmand/gateway/src/devmand/channels/cli/engine/Engine.cpp (hunk 4 above)
Comment (tomasol) [reply devmand/gateway/src/devmand/channels/cli/engine/Engine.cpp]: Done

File: devmand/gateway/src/devmand/channels/cli/engine/Engine.cpp (hunk 3 above)
Comment (tomasol) [reply devmand/gateway/src/devmand/channels/cli/engine/Engine.cpp]: Reworded TODO

File: devmand/gateway/src/devmand/channels/cli/engine/Engine.cpp (hunk 2 above)
Comment (tomasol) [reply devmand/gateway/src/devmand/channels/cli/engine/Engine.cpp]: Done, thanks

File: devmand/gateway/CMakeLists.txt (hunk 1 above)
Comment (tomasol) [reply devmand/gateway/CMakeLists.txt:164]: Let's move the files once WriterRegistry and dependent PRs are finished to avoid possible conflicts.

**Review by KenG98 — APPROVED** [review]

**Pull request closed without merging**